File: frontend/src/lib/complexity.ts
import type { StrategySpec } from '../types'

export type ComplexityLabel = 'Low' | 'Medium' | 'High'

export interface Complexity {
  score: number
  label: ComplexityLabel
}

// DESIGN.md §3: "more nodes and more parameters means more overfitting
// risk." A simple, honest count — not a real holdout gate (Lab/M7 doesn't
// exist yet), just a nudge.
export function computeComplexity(spec: StrategySpec): Complexity {
  const score = spec.nodes.reduce((total, node) => total + 1 + Object.keys(node.params ?? {}).length, 0)
  const label: ComplexityLabel = score < 8 ? 'Low' : score < 15 ? 'Medium' : 'High'
  return { score, label }
}

export const COMPLEXITY_CAVEAT =
  'More nodes and parameters mean more ways to accidentally fit noise instead of a real pattern. ' +
  'A backtest alone can\'t tell the difference — that\'s what honest testing (coming soon) is for.'

File: backend/engine/complexity.py
"""Direct Python port of frontend/src/lib/complexity.ts's formula — needed
here because the copilot's `create_strategy`/`update_strategy` tool results
(backend/ai/tools.py) can't reach into frontend TS. Keep the two in sync by
hand; the formula is small and stable (DESIGN.md §3: "more nodes and more
parameters means more overfitting risk" — a simple, honest count, not a
real holdout gate)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from backend.engine.graph.spec import StrategySpec

ComplexityLabel = Literal["Low", "Medium", "High"]


@dataclass(frozen=True)
class Complexity:
    score: int
    label: ComplexityLabel


def compute_complexity(spec: StrategySpec) -> Complexity:
    score = sum(1 + len(node.params) for node in spec.nodes)
    label: ComplexityLabel = "Low" if score < 8 else "Medium" if score < 15 else "High"
    return Complexity(score=score, label=label)
